js-packages/@prestojs/rest/src/PageNumberPaginator.ts

```diff
@@ -0,0 +1,182 @@
+        }
+        const nextState = { ...this.currentState, pageSize };
+        if (pageSize == null) {
+            delete nextState.page;
+            delete nextState.pageSize;
+        }
+        return nextState as PageNumberPaginationState;
+    }
+
+    setPageSize(pageSize: null | number): void {
+        this.setCurrentState(this.pageSizeState(pageSize));
+    }
+
+    /**
+     * Return the state for the next page
+     *
+     * Does not transition state. To transition state call `next` instead.
+     */
+    nextState(): PageNumberPaginationState {
+        return this.pageState((this.page || 1) + 1);
+    }
+
+    /**
+     * Go to the next page.
+     */
+    next(): void {
+        this.setCurrentState(this.nextState());
+    }
+
+    /**
+     * Return the state for the previous page
+     *
+     * Does not transition state. To transition state call `previous` instead.
+     */
+    previousState(): PageNumberPaginationState {
+        return this.pageState((this.page || 1) - 1);
+    }
+
+    /**
+     * Go to the previous page.
+     */
+    previous(): void {
+        this.setCurrentState(this.previousState());
+    }
+
+    /**
+     * Return the state for the first page
+     *
+     * Does not transition state. To transition state call `first` instead.
+     */
+    firstState(): PageNumberPaginationState {
+        return this.pageState(1);
+    }
+
+    /**
+     * Go to the first page.
+     */
+    first(): void {
+        this.setCurrentState(this.firstState());
+    }
+
+    /**
+     * Return the state for the first page
+     *
+     * Does not transition state. To transition state call `last` instead.
+     */
+    lastState(): PageNumberPaginationState | null {
+        if (null == this.total || null == this.pageSize) {
+            return null;
+        }
+        return this.pageState(Math.ceil(this.total / this.pageSize));
+    }
+
+    /**
+     * Go to the last page. If the last page isn't yet known (eg. results
+     * haven't yet been returned) then null will be returned.
+     *
+     * If the last page is not yet known because results haven't been returned this function
+     * does nothing.
+     */
+    last(): void {
+        const nextState = this.lastState();
+        if (nextState) {
+            this.setCurrentState(nextState);
+        }
+    }
+
+    getRequestInit({ query, ...options }): EndpointExecuteOptions {
+        const newQuery = { ...query };
+        if (this.currentState.pageSize) {
+            newQuery.pageSize = this.currentState.pageSize;
+        }
+        if (this.currentState.page) {
+            newQuery.page = this.currentState.page;
+        }
+        return {
+            query: newQuery,
+            ...options,
+        };
+    }
+
+    setResponse({ total, pageSize }: { total: number; pageSize?: number }): void {
+        this.setInternalState({ total });
+        if (pageSize && this.currentState.pageSize !== pageSize) {
+            this.setPageSize(pageSize);
```

Comment:
Out of curiosity, why do you need to check if the page size has changed? Can't you just set it and react will decide nothing needs to be changed (some overhead, but presumably not a lot), or does setting state trigger a rerender even if nothing changed?

Reply:
setState doesn't do deep equality so will re-render as state is stored as object. I don't think it really matters though - I could have left that check out